File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/TaskProviderWrapper.java
package com.example.alexeyglushkov.taskmanager.task;

import android.os.Handler;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by alexeyglushkov on 11.06.17.
 */

public class TaskProviderWrapper implements TaskProvider {

    private TaskProvider provider;
    private Map<TaskPoolListener, TaskPoolListener> listenerMap = new HashMap<>();

    public TaskProviderWrapper(TaskProvider provider) {
        this.provider = provider;
    }

    @Override
    public void setHandler(Handler handler) {
        provider.setHandler(handler);
    }

    @Override
    public Handler getHandler() {
        return provider.getHandler();
    }

    @Override
    public void setTaskProviderId(String id) {
        provider.setTaskProviderId(id);
    }

    @Override
    public void addTask(Task task) {
        provider.addTask(task);
    }

    @Override
    public void removeTask(Task task) {
        provider.removeTask(task);
    }

    @Override
    public String getTaskProviderId() {
        return provider.getTaskProviderId();
    }

    @Override
    public Task getTask(String taskId) {
        return provider.getTask(taskId);
    }

    @Override
    public int getTaskCount() {
        return provider.getTaskCount();
    }

    @Override
    public List<Task> getTasks() {
        return provider.getTasks();
    }

    @Override
    public void setUserData(Object data) {
        provider.setUserData(data);
    }

    @Override
    public Object getUserData() {
        return provider.getUserData();
    }

    @Override
    public void addListener(final TaskPoolListener listener) {
        TaskPoolListener wrapperListener = new TaskPoolListener() {
            @Override
            public void onTaskAdded(TaskPool pool, Task task) {
                listener.onTaskAdded(TaskProviderWrapper.this, task);
            }

            @Override
            public void onTaskRemoved(TaskPool pool, Task task) {
                listener.onTaskRemoved(TaskProviderWrapper.this, task);
            }
        };

        listenerMap.put(listener, wrapperListener);
        provider.addListener(wrapperListener);
    }

    @Override
    public void setPriority(int priority) {
        provider.setPriority(priority);
    }

    @Override
    public void removeListener(TaskPoolListener listener) {
        TaskPoolListener wrappedListener = listenerMap.get(listener);
        if (wrappedListener != null) {
            provider.removeListener(wrappedListener);
            listenerMap.remove(listener);
        }
    }

    @Override
    public int getPriority() {
        return provider.getPriority();
    }

    @Override
    public Task getTopTask(List<Integer> typesToFilter) {
        return provider.getTopTask(typesToFilter);
    }

    @Override
    public Task takeTopTask(List<Integer> typesToFilter) {
        return provider.takeTopTask(typesToFilter);
    }

    @Override
    public void onTaskStatusChanged(Task task, Task.Status oldStatus, Task.Status newStatus) {
        provider.onTaskStatusChanged(task, oldStatus, newStatus);
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/TaskProviderWrapper.kt
package com.example.alexeyglushkov.taskmanager.task

import android.os.Handler
import androidx.annotation.WorkerThread
import kotlinx.coroutines.CoroutineScope

import java.util.HashMap

/**
 * Created by alexeyglushkov on 11.06.17.
 */

open class TaskProviderWrapper(val provider: TaskProvider) : TaskProvider {
    private val listenerMap = HashMap<TaskPool.Listener, TaskPool.Listener>()
    override var taskFilter: TaskProvider.TaskFilter?
        get() = provider.taskFilter
        set(value) {
            provider.taskFilter = value
        }

    override var threadRunner: ThreadRunner
        get() = provider.threadRunner
        set(handler) {
            provider.threadRunner = threadRunner
        }

    override var taskProviderId: String
        get() = provider.taskProviderId
        set(id) {
            provider.taskProviderId = id
        }

    override fun getTaskCount(): Int {
        return provider.getTaskCount()
    }

    override fun getTasks(): List<Task> {
        return provider.getTasks()
    }

    override var userData: Any?
        get() = provider.userData
        set(data) {
            provider.userData = data
        }

    override var priority: Int
        get() = provider.priority
        set(priority) {
            provider.priority = priority
        }

    override fun addTask(task: Task) {
        provider.addTask(task)
    }

    override fun removeTask(task: Task) {
        provider.removeTask(task)
    }

    override fun cancelTask(task: Task, info: Any?) {
        provider.cancelTask(task, info)
    }

    override fun getTask(taskId: String): Task? {
        return provider.getTask(taskId)
    }

    override fun addListener(listener: TaskPool.Listener) {
        val wrapperListener = object : TaskPool.Listener {
            override fun onTaskAdded(pool: TaskPool, task: Task) {
                listener.onTaskAdded(this@TaskProviderWrapper, task)
            }

            override fun onTaskRemoved(pool: TaskPool, task: Task) {
                listener.onTaskRemoved(this@TaskProviderWrapper, task)
            }

            override fun onTaskCancelled(pool: TaskPool, task: Task, info: Any?) {
                listener.onTaskCancelled(this@TaskProviderWrapper, task, info)
            }

            override fun onTaskConflict(pool: TaskPool, newTask: Task, oldTask: Task): Task {
                return listener.onTaskConflict(this@TaskProviderWrapper, newTask, oldTask)
            }
        }

        listenerMap[listener] = wrapperListener
        provider.addListener(wrapperListener)
    }

    override fun removeListener(listener: TaskPool.Listener) {
        val wrappedListener = listenerMap[listener]
        if (wrappedListener != null) {
            provider.removeListener(wrappedListener)
            listenerMap.remove(listener)
        }
    }

    @WorkerThread
    override fun getTopTask(): Task? {
        return provider.getTopTask()
    }

    @WorkerThread
    override fun takeTopTask(): Task? {
        return provider.takeTopTask()
    }

    override fun onTaskStatusChanged(task: Task, oldStatus: Task.Status, newStatus: Task.Status) {
        provider.onTaskStatusChanged(task, oldStatus, newStatus)
    }
}
